design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+
+- Users no longer have to select their full architecture up front as with the current ECS patterns. Instead of chosing between `ApplicationLoadBalancedEc2Service` and `NetworkLoadBalancedEc2Service` up front, they can actually deploy a service, and then later add or remove an `HttpLoadBalancerAddon` from the service without needing to change the base class in their code.
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
+
+## Service Lifecycle
+
+The new `Service` construct differs a bit from other CDK L2+ constructs. Most CDK constructs are fully assembled in the constructor and are then immutable. The `Service` class is not like this. The constructor does nothing but initialize an empty list of extensions. Users add extensions to this list, to define their intention about what should be built later on. The service constructs do not actually exist at this point. The only thing that exists is a list of user specified intentions about what constructs they do want to exist.
+
+The service's constructs are instantiated once the service is added to an environment. At that point all the extensions are executed to build out the actual CDK resources. From this point on the service constructs are available and immutable.
+
+## Extension Implemention
+
+Each extension is a collection of functions which are executed by the `Service` class when the service is added to an environment. Functions are executed in the order that the extensions were added to the service. Some functions are mutating hooks which modify service or task definition properties prior to these constructs being created. Other functions are called after each construct has been created so that the extension can make use of the construct.
+
+The hooks for an extension are as follows:
+
+- `addHooks()` - This hook is called after all the extensions are added, but before anything else has been run. It gives each extension a chance to do some inspection of the overall service and see what other extensions have been added. Some addons may want to register hooks on the other addons to modify them. For example the Firelens addon wants to be able to modify the settings of the application container to route logs through Firelens.
+- `mutateTaskDefinitionProps()` - This is a mutating hook which is passed the `ecs.TaskDefinitionProps` for a TaskDefinition. This allows the extension to make modifications to the task definition, such as App Mesh extension modifying the proxy settings for the task.
```

Comment:
I don't love the verb `mutate` here -- has the Task Definition even been created yet? Would `override` or `modify` work better?

Reply:
Sure thing I don't have any issues with changing it to "modify". To be clear this hook only modifies the `props` for the `TaskDefinition` that is about to be created. The `TaskDefinition` has not yet been created at this point, there is just a proposed `props` for it. Each extension is given the option of changing the `props` and then once they have all changed the `props` to fit their needs the resulting `props` object is passed to the `TaskDefinition` constructor to create a new `TaskDefinition`. I'll try to clarify this behavior in the doc